src/expr/src/vector_op/bitwise_op.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::any::type_name;
+use std::convert::TryInto;
+use std::fmt::Debug;
+use std::ops::{BitAnd, BitOr, BitXor, Not};
+
+use num_traits::{CheckedShl, CheckedShr};
+use risingwave_common::error::ErrorCode::{InternalError, NumericValueOutOfRange};
+use risingwave_common::error::{Result, RwError};
+
+use crate::vector_op::arithmetic_op::general_atm;
+
+// Conscious decision for shl and shr is made here to diverge from PostgreSQL.
+// If overflow happens, instead of truncated to zero, we return overflow error as this is unexpected
+// behaviour If the RHS is negative, instead of having an unexpected answer, we return an error.
+#[inline(always)]
+pub fn general_shl<T1, T2>(l: T1, r: T2) -> Result<T1>
+where
+    T1: CheckedShl + Debug,
+    T2: TryInto<u32> + Debug,
+{
+    general_shift(l, r, |a, b| match a.checked_shl(b) {
+        Some(c) => Ok(c),
+        None => Err(RwError::from(NumericValueOutOfRange)),
+    })
```

Comment:
FYI `a.checked_shl(b).ok_or_else`
Not sure if this looks better or worse.

Reply:
Alright I change it for mine part.  `general_shift(l, r, |a,b| a.checked_shl(b).ok_or_else( || RwError::from(NumericValueOutOfRange)))`. I feel it looks better.

Just to point out, the previous writing I was copying from `arithmetic_op.rs`. Should it be changed as well for consistency across similar code?

Reply:
Keeping them consistent would be better.